File: divideAndConquer/python/sum.py
"""
Grokking Algorithms Ch 4 Exercises
Divide and Conquer
"""

## write out the code for the recursive sum D&C function
import collections

def sum(array) :
  result = 0
  if (len(array)==0):
    return 0
  else:
    removal = array[0]
    array.remove(removal)
    return result + removal + sum(array)


print(sum([1,2,3]))
print(sum([4,5,6]))

File: divideAndConquer/sum.js
/*
Grokking Algorithms Ch 4 Exercises
Divide and Conquer
*/
// write out the code for the recursive sum D&C function
const sum = (array) => {
  let result = 0;
  // recursive case
  if (array.length) {
    let newAdd = array.shift();
    return result + newAdd + sum(array);
  // base case
  } else {
    return 0;
  }
};
export default sum;